plugin/trino-postgresql/src/test/java/io/trino/plugin/postgresql/TestPostgreSqlConnectorTest.java

```diff
@@ -1013,6 +1017,68 @@ public void testInPredicatePushdown()
+            assertThat(query(session, "SELECT a.a2, b.b2 FROM %s a JOIN %s b ON CAST(a.a1 + 123 AS %s) = b.b1".formatted(tableA.getName(), tableB.getName(), testCase.castType())))
+                    .isFullyPushedDown();
+        }
+
+        // cast with pushdownable complex expression with function
+        testCase = new CastTestCase("varchar", "tinyint", "smallint");
+        try (TestTable tableA = new TestTable(onRemoteDatabase(), "test_join_cast_a_", "(a1 %s, a2 %s)".formatted(testCase.fromRemoteType(), testCase.fromRemoteType()));
+                TestTable tableB = new TestTable(onRemoteDatabase(), "test_join_cast_b_", "(b1 %s, b2 %s)".formatted(testCase.targetRemoteType(), testCase.targetRemoteType()))) {
+            assertThat(query(session, "SELECT a.a2, b.b2 FROM %s a JOIN %s b ON CAST(UPPER(a.a1) AS %s) = b.b1".formatted(tableA.getName(), tableB.getName(), testCase.castType())))
+                    .isFullyPushedDown();
+        }
+
+        // cast with non-pushdownable complex expression
+        testCase = new CastTestCase("date", "varchar");
+        try (TestTable tableA = new TestTable(onRemoteDatabase(), "test_join_cast_a_", "(a1 %s, a2 %s)".formatted(testCase.fromRemoteType(), testCase.fromRemoteType()));
+                TestTable tableB = new TestTable(onRemoteDatabase(), "test_join_cast_b_", "(b1 %s, b2 %s)".formatted(testCase.targetRemoteType(), testCase.targetRemoteType()))) {
+            assertThat(query(session, "SELECT a.a2, b.b2 FROM %s a JOIN %s b ON CAST(year(a.a1) AS %s) = b.b1".formatted(tableA.getName(), tableB.getName(), testCase.castType())))
+                    .joinIsNotFullyPushedDown();
+        }
+
+        // cast column/expression where native type was lost during read mapping from one side, and from another column which was not read and type transformed by trino because of occurred pushdown
+        CastTestCase testCaseMoney = new CastTestCase("money", "varchar(100)", "money");
+        try (TestTable tableA = new TestTable(onRemoteDatabase(), "test_join_cast_a_", "(a1 %s, a2 %s)".formatted(testCaseMoney.fromRemoteType(), testCaseMoney.fromRemoteType()));
+                TestTable tableB = new TestTable(onRemoteDatabase(), "test_join_cast_b_", "(b1 %s, b2 %s)".formatted(testCaseMoney.targetRemoteType(), testCaseMoney.targetRemoteType()))) {
+            assertThatThrownBy(() -> getQueryRunner().execute(session, "SELECT a.a2, b.b2 FROM %s a JOIN %s b ON CAST(a.a1 AS %s) = b.b1".formatted(tableA.getName(), tableB.getName(), testCaseMoney.castType())));
+            // the best we can do - cast right side to a trino type, to which is resolved left side. After pushdown, trino type is mapped to native type
+            assertThat(query(session, "SELECT a.a2, b.b2 FROM %s a JOIN %s b ON CAST(a.a1 AS %s) = CAST(b.b1 AS %s)".formatted(tableA.getName(), tableB.getName(), testCaseMoney.castType(), testCaseMoney.castType())))
```

Comment:
This is a breaking change, so has to be covered by config/session property toggle.
@Praveen2112